noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,309 @@
+    #[aztec(view)]
+    fn admin() -> Field {
+        storage.admin.read().to_field()
+    }
+
+    #[aztec(public)]
+    #[aztec(view)]
+    fn is_minter(minter: AztecAddress) -> bool {
+        storage.minters.at(minter).read()
+    }
+
+    #[aztec(public)]
+    fn transfer_public(from: AztecAddress, to: AztecAddress, token_id: Field, nonce: Field) {
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let public_owner = storage.public_owners.at(token_id).read();
+        assert(public_owner.eq(from), "invalid owner");
+
+        storage.public_owners.at(token_id).write(to);
+    }
+
+    // We have only public burn flow because the flow is not that common for it to be worth it to clutter
+    // the contract with a private one.
+    #[aztec(public)]
+    fn burn(from: AztecAddress, token_id: Field, nonce: Field) {
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit_public(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let public_owner = storage.public_owners.at(token_id).read();
+        assert(public_owner.eq(from), "invalid owner");
+
+        storage.nft_exists.at(token_id).write(false);
+        storage.public_owners.at(token_id).write(AztecAddress::zero());
+    }
+
+    /// Prepares a shield partial note which can later on be finalized by the `finalizer`.
+    // TODO(#8238): Remove the `note_randomness` argument below once we have partial notes delivery (then we can just
+    // fetch the randomness from oracle).
+    #[aztec(private)]
+    fn prepare_shield(
+        finalizer: AztecAddress,
+        note_randomness: Field,
+        transient_storage_slot_randomness: Field
+    ) {
+        // We create a partial NFT note with unpopulated/zero token id for msg_sender
+        let shield_owner = context.msg_sender();
+        let shield_owner_npk_m_hash = get_current_public_keys(&mut context, shield_owner).npk_m.hash();
+        let shield_note_slot = storage.private_nfts.at(shield_owner).storage_slot;
+        let hiding_point = NFTNoteHidingPoint::new(shield_owner_npk_m_hash, shield_note_slot, note_randomness);
+
+        // We make the owner part of the slot preimage to ensure he cannot interfere with non-owner's slots
+        let owner_storage_slot_commitment: Field = pedersen_hash(
+            [shield_owner.to_field(), transient_storage_slot_randomness],
+            TRANSIENT_STORAGE_SLOT_PEDERSEN_INDEX
+        );
+        // Then we hash the owner storage slot commitment with finalizer and use that as the final shield slot
+        // --> by hashing it with a finalizer we ensure that finalizer cannot interfere with slots not assigned to him.
+        let slot: Field = pedersen_hash(
+            [finalizer.to_field(), owner_storage_slot_commitment],
```

Comment:
> We need to pass the randomness as a param so that we can derive the slot before making the call and pass owner_storage_slot_commitment in the second batched call, correct?

That's correct. Theoretically we could call `unsafe_rand` oracle in this function, then return the value and make the return value be passed as arg to the `send_to_shield` function but that is not supported by current account contract implementation (and I think it would be quite a large refactor because the account contracts expect the calls to be fully formed in the entry point payloads).

> This would be much simpler if we could have a stack or a queue of transient notes

Interesting. But how do you imagine that ordering would be preserved? Would you have the expectation that it's up to the developer to call `prepare_shield` and `send_to_shield` in the correct order if dealing with multiple concurrent shields? (I guess this is fine, I have just not really though about this)

> But as a side bonus we'd support doing multiple shields at once

This is supported by the current code as well as you can control that via `transient_storage_slot_randomness` function arg. But yes, the devex is 💩

Reply:
We discusssed today that we'd follow this approach for now and look into improvements later.